docs/src/adr/018adr-inlining.md

```diff
@@ -0,0 +1,156 @@
+
+More precisely, suppose we are given a _non-recursive_ operator `A` with the following definition:
+```tla
+A(p1,...,pn) == body
+```
+
+The term "inlining" (of `A`) typically refers to the process of replacing instances of operator application `A(e1,...,en)` with `body[e1/p1,...,en/pn]`, i.e. the expression obtained by replacing each instance of `pi` with `ei` within `body`.
+We elect to use the term in a broader sense of "replacing an operator with its definition", and define two "flavors" of inlining:
+  1. Standard inlining: the instantiation described above
+      1. Non-nullary inlining: the instantiation described above, except the inlining skips nullary LET-defined operators
+  2. Pass-by-name inlining: replacing an operator name `A` used as an argument to a higher-order operator with a local LET-definition: `LET A_LOCAL(p1,...,pn) == A(p1,...,pn) IN A_LOCAL`
+
+The reason for doing (1) is that, at some point, a rewriting rule would have to generate constraints from `A(e1,...,en)`. 
+To do this, we couldn't just separately encode `body` and `e1,...,en`, because the richness of the data structures allowed in TLA+ makes it difficult to combine independently generated constraints, in cases where the operator parameters are complex expressions (e.g. `e1` is some record with varied nested constructs).
+This is mostly due to the fact that there is no 1-to-1 correspondence between TLA+ objects and SMT datatypes, so encoding object equality is more complicated (which would be needed to express that `ei` instantiates `pi`).
+Therefore we must, no later than at the point of the rewriting rule, know `body[e1/p1,...,en/pn]`. 
+
+While inlining non-nullary operators is  strictly necessary, inlining nullary operators is not, because nullary operators, by definition, do not have formal parameters. 
+Therefore, in a well-constructed expression, all variables appearing in a nullary operator are scoped, i.e. they are  either specification-level variables (defined as `VARIABLE`), or bound in the context within which the operator is defined, if local. An example of the latter would be `i` in 
+```tla
+\E i \in S: LET i2 == i * i in i2 = 0
+```
+which is not bound in the nullary operator `i2`, but it is defined in the scope of the `\E` operator, under which `i2` is defined. Therefore, any analysis of `i2` will have `i` in its scope.
+The non-nullary variant of (1) is therefore strictly better for performance, because it allows for a sort of caching, which avoids repetition. Consider for example:
+```tla
+A1(p) == 
+  LET pCached == p
+  IN F(pCached,pCached)
+
+A2(p) == F(p,p)
+```
+
+If we apply the substitution `{p -> e}`, for some complex value `e`, to the bodies of both operators, the resutls are 
+
+```tla
+LET pCached == e
+IN F(pCached,pCached)
+```
+and
+```
+F(e,e)
+```
+
+In the first case, we can translate `pCached` to a cell and reuse the cell expression twice, whereas in the second, `e` is rewritten twice, independently.
```

Comment:
Maybe link

> translate `pCached` to a cell

to the OOPSLA paper (or another source)? Cells weren't introduced in the scope of this ADR.